Reword and add uniformity to Standard Library documentation (#4464)
edgedb/edgedb: docs/stdlib/array.rst

Constructing arrays
^^^^^^^^^^^^^^^^^^^

> Array constructors are an expression that allow for a sequence of

Reviewer: ```suggestion
Arrays are constructed by placing multiple comma-separated expressions within
square brackets. Here are a few examples:
```

> comma-separated expressions *of the same type* enclosed in square brackets.

Reviewer: ```suggestion
```

> This will produce an array of the following:

Author: Not sure if it helps too much to add clarification on it after constructors are mentioned above this. I can go ahead and add this though!

> This will produce an array of the following:

Reviewer: This could just be a misunderstanding on my part of the synopsis block below (and maybe _that's_ where the actual change needs to happen; I don't know). What is the synopsis block below supposed to represent? I thought it was demonstrating the syntax of an array constructor to the user, hence my suggestion. We just need to make sure the text leading up to that colon is introducing what comes after it.

> This will produce an array of the following:

Author: I think when it's read altogether it reads out as "this [constructor] will produce an array," and goes in reference to the details above it

> .. eql:synopsis::

Reviewer: Recommend removing this since the grammar is difficult to parse and is conveyed via the examples.

Just for added clarity, the full recommendation would be to have these lines:
> Arrays are constructed by placing multiple comma-separated expressions within
> square brackets. Here are a few examples:

followed by the array examples that currently start on line 70 with no other content between.

    "[" <expr> [, ...] "]"

> You can then use these arrays in EdgeDB to access information.

Author: I'm not quite sure if I'm following. "Access" is in reference to `SELECT` being used in the code block below. Should we be clarifying that the example is about accessing a set array property?

> You can then use these arrays in EdgeDB to access information.

Reviewer: Yes, but you're saying the array is being used to access information which doesn't ring true to me. The select is being used to access information, but the array is just the construct being used to store the data being accessed. The array itself is not doing any accessing of information.

> You can then use these arrays in EdgeDB to access information.

Reviewer: ```suggestion
```

Empty arrays
^^^^^^^^^^^^

> An empty array can also be created, but it must be used with

Reviewer: ```suggestion
You can also create an empty array, but it must be done by providing the type
information using type casting. EdgeDB cannot infer the type of an empty array
created otherwise. Example:
```

> a type cast as EdgeDB cannot infer the type of an array with no elements:

Reviewer: ```suggestion
```

    :index: array

    Represents a one-dimensional array of an homogeneous ordered list.

>     Array indexing will always start at zero. With the exception of other

Reviewer: One more recommendation here that came out of my review session with @vpetrovykh:

Keep these two sentences as separate paragraphs since they are unrelated.

Also, add another one:

> In EdgeDB, arrays are always one-dimensional.

This isn't critical to the one-liner and can be relayed in the details instead.

    array types, any type may be used as the given element contained within.

    Array types are implicitly created from :ref:`array
    constructors <ref_std_array_constructor>` as seen here:

    .. code-block:: edgeql-repl

        db> select [1, 2];
        {[1, 2]}

>     The syntax of an array type's declaration can be found from :ref:`this

Reviewer: Avoid links like "this section," "click here," and similar

Author: Most likely missed this one while cleaning the references. Good catch!

Reviewer: Instead of reworking this link, maybe we could just provide an example of an array declaration inline. We can use one similar to the page we're currently linking out to.

Author: This also works

    section <ref_datamodel_arrays>`.

    Please see also the list of standard :ref:`array

>     functions <ref_std_array>`, as well as generic functions such as

Reviewer: We should link to the list of generic functions too.

Author: Agreed, `ref_stdlib_generic` should work perfect for this

.. eql:operator:: arrayidx: array<anytype> [ int64 ] -> anytype

>     Indexes an array of :eql:type:`anytype`.

Reviewer: I'm not sure this provides much clarity. Suggestion:
> Accesses the element of the array at the given index (zero-indexed).

I think with that, you could maybe even drop the next sentence since the description will already tell you you're going to get back the thing at the index you specified.

Author: The main problem I run into with trying to simplify everything on the first paragraph is the 79 character limit. I'll likely have to reword the first paragraph to follown something similar, otherwise this will have to be the 2nd paragraph.

Reviewer: Can you use the line continuation character to go beyond 79 characters if you need to? I'm pretty new to rST/Sphinx, so forgive me if that's not possible.

If you end up needing something shorter, this might work just as well:

> Accesses the array element at the index (zero-indexed)

>     Indexes an array of :eql:type:`anytype`.

Reviewer: ```suggestion
    Accesses the element of the array at a given index.
    
    Arrays in EdgeDB are zero-indexed.
```

Clarify explanation and demote zero-indexed to the description.

Author: Can we italicize "array" as what it's named in the argument signature or re-word it to use "an" instead? "the array" sounds a little jarring to me

>     This results in a representable reference of the array element specified:

Reviewer: Revert this since all we're looking to do here is introduce the next example.

        db> select [(x := 1, y := 1), (x := 2, y := 3.3)][1];
        {(x := 2, y := 3.3)}

>     This operator also allows accessing elements through negation.

Reviewer: Would be nice to briefly explain how this works for a user who might not know. Suggestion:
> This operator also allows accessing elements from the end of the array using a negative index.

        db> select [(x := 1, y := 1), (x := 2, y := 3.3)][1];
        {(x := 2, y := 3.3)}

>     This operator may also be used when trying to negatively index elements.

Reviewer 2: I don't like the "trying" word here. There's no trying - you just index :)

Author: Will change

    .. code-block:: edgeql-repl

        db> select [1, 2, 3][-1];
        {3}

>     However, referencing an non-existent element of an array will result in

Reviewer 2: ```suggestion
    However, referencing a non-existent element of an array will result in
```
?

Reviewer 2: I'm also not sure why "However" is used here; non-existing element error is really not connected in any way to the negative indexing.

.. eql:operator:: arrayslice: array<anytype> [ int64 : int64 ] -> anytype

>     Slices an array of :eql:type:`anytype` containing :eql:type:`int64`.

Reviewer: Let's try to explain slicing briefly. Suggestion:
> Produces an array containing a subset of the original array's elements.

This may not be 100% accurate since you could slice an array with all of the original's elements, but I would like to find a way to tell what the operator does rather than repeating the name of it. Open to suggestions.

Author: I'll give this a go, ideally I want to make a reference to that type and the character limit may pose an issue with that

>     Slices an array of :eql:type:`anytype` containing :eql:type:`int64`.

Reviewer: ```suggestion
    Produces a sub-array from an existing array.
```

New recommendation to avoid "set" outside the context of an bona fide set.

    This results in a representable reference of the array's elements.

>     Omitting the lower bound an array will default the result to zero.

Reviewer: >... will default the result to zero.
This reads as though the result of the slice will default to zero when actually the low bound will default to zero.

Suggestion:
> ... will result in a default low bound of zero.

Reviewer: ```suggestion
    Omitting the lower bound of an array slice will default to a low bound of
    zero.
```

>     Doing so to the upper bound will also default to the current size of the

Reviewer: Suggestion:
> Doing so to the upper bound will ~also~ default to the current size of the...

Reviewer: ```suggestion
    Omitting the upper bound will also default to the length of the array. The
```

>     array. The upper bound of an array is non-inclusive:

Reviewer: ```suggestion
    upper bound of an array is non-inclusive:
```

>     array. The upper bound of an array is non-inclusive:

Reviewer: ```suggestion
    array.
    
    The lower bound of an array slice is inclusive while the upper bound is
    not.
```

.. eql:operator:: arrayslice: array<anytype> [ int64 : int64 ] -> anytype

    Produces a sub-array from an existing array.

>     This results in a representable reference of the array's elements.

Reviewer 2: I've reread this sentence a few times and I can't understand what it says.

    Omitting the lower bound of an array slice will default to a low bound of
    zero.
    Omitting the upper bound will default the upper bound to the length of the
    array.
    
    The lower bound of an array slice is inclusive while the upper bound is

>     not.

Reviewer 2: Perhaps this should be formatted as bullets

.. eql:operator:: arrayslice: array<anytype> [ int64 : int64 ] -> anytype

    Slices an array of :eql:type:`anytype` containing a fixed range of

>     :eql:type:`int64` into a representable reference of the array's elements.

Reviewer 2: This will be an error, mush be a single sentence/paragraph fitting in 79 characters.

    Omitting the lower bound an array will default the result to zero.
    Doing so to the upper bound will also default to the current size of the
    array.

>     The upper bound an array is non-inclusive.

Reviewer 2: I can't parse the "an array" part.

Author: Typo, thank you for catching, should be "... of an array" or "... of the array"

        db> select [1, 2, 3][:-2];
        {[1]}

>     Referencing an array slice outside of the array's boundaries will result

Reviewer: Suggestion:
> Slicing an array using values for both upper and lower bounds that are outside the array's boundaries will result in an empty array.

Reviewer: ```suggestion
    If your array slice boundaries do not include any valid index from the
    array, the slice will produce an empty array. Slicing with a lower bound
    less than the minimum index or a upper bound greater than the
    maximum index are functionally equivalent to not specifying those bounds
    for your slice.
```

>     in an empty array, unlike accessing an element with a direct index:

Reviewer: ```suggestion
```

        db> select [1, 2, 3][:-2];
        {[1]}

>     If your array slice boundaries do not include any valid index from the

Reviewer 2: re "your": I don't quite like this style of referring to the reader. Do we have to? IMO the previous wording is succinct and fine.

.. eql:operator:: arrayplus: array<anytype> ++ array<anytype> -> array<anytype>

>     Concatenates given arrays of :eql:type:`anytype` into one.

Reviewer: ```suggestion
    Concatenates two arrays of the same type into one.
```

Important to note that the types must be the same.

Author: I think we'll need to mirror this in a few more places.. Good catch!

>     This results in a reference of both array's elements conjoined together:

Reviewer: This seems like it would be easy to misinterpret. Maybe this instead?
> This results in an array containing the elements of both of the concatenated arrays.

    :index: aggregate array set

    Returns an array made from all of the input set elements.

>     The ordering of the inputted set will be preserved if specified:

Reviewer: Recommend leaving this as "input set" for consistency with line 214.

    :index: set array unpack

    Returns the elements of an array as a set:

    .. code-block:: edgeql-repl

        db> select array_unpack([2, 3, 5]);
        {3, 2, 5}

>     .. note::

Reviewer: Completely missed this. This change should be reverted because the example above illustrates the information referenced in the note. We should give the note first to give the context beforehand.

Once you revert it, you will also want to remove the colon from the one-liner.

Author: I believe I did this as note admonitions were sometimes split between the top and bottom of it. In hindsight we should remove the admonition in whole and have it as a separate paragraph, it's objectively a fundamental aspect of the function now that I see it.

    :index: join array_to_string implode

>     Returns the elements of an ``array`` joined together with a ``delimiter``:

Reviewer: Suggestion:
> Returns the elements of \`\`array\`\` as a string, joined together with a ``delimiter``.

- Removed "an" before \`\`array\`\` since we know the array being operated on.
- We should specify the result is a string.

Author: This suggestion will not work as it exceeds 79 chars on the first paragraph

Reviewer: How about this?
> Returns ``array`` elements as a string joined by ``delimiter``

    :index: join array_to_string implode

>     Returns the elements of an array joined as a string by a delimiter.

Reviewer 2: I like the previous wording here, IMO it's easier to read and understand. I've troubles parsing the new version fast.

    :index: fill

>     Returns a new array of size ``n`` with the specified ``val``:

Reviewer: Since the one-liners are recycled in many cases for the tables at the top, we want to globally ensure none of them ends with a colon so that it doesn't make it into that table. A period, even before the code sample would be fine.

Reviewer: Learned from @vpetrovykh that we have a style of styling parameters by surrounding them with `*`. Also, since they will also be rendered outside the context of this (without the function signature) using `n` and `val` here will lose meaning.

```suggestion
    Returns a new array of a specified number of copies of the specified value.
```

Author: To avoid an aggressive amount of suggested changes, I'll go ahead and look through all of our functions and get these reverted, thank you for the heads up. I chose to code-format it instead of `*` so it'd look similar to the way the client renders it. Would it be possible if we could have the original syntax color format it as green like it is in the argument signature so it stands out a little better? I think this would be a great benefit

Reviewer: @i0bs You're asking if emphasis (i.e., `*something*`) can be styled in green so that those references will be a consistent color? Confirming my understanding.

Author: Yep! Maybe if we color code to the format we can make it work better with the style we're wanting to follow here. Since we're doing this to reference arguments in the signature, we should try and respect that style heavily. Italics alone can be mixed with other italicized comments

    :index: fill

>     Returns an array of specified size, filled with specified value.

Reviewer 2: grammar

```suggestion
    Returns an array of the specified size, filled with the provided value.
```
